sqlmesh/core/engine_adapter/base.py

```diff
@@ -199,9 +204,18 @@ def catalog_support(self) -> CatalogSupport:
         return CatalogSupport.UNSUPPORTED
 
     @classmethod
-    def _casted_columns(cls, columns_to_types: t.Dict[str, exp.DataType]) -> t.List[exp.Alias]:
+    def _casted_columns(
+        cls,
+        columns_to_types: t.Dict[str, exp.DataType],
+        source_columns: t.Optional[t.List[str]] = None,
+    ) -> t.List[exp.Alias]:
+        source_columns = source_columns or list(columns_to_types)
```

Comment:
should this be a set? I see that each column gets looked up

Reply:
I originally used a set but it creates a bit of a trap for developers: It can be very tempting, if needing to select columns from a user query/df, to use `source_columns`. If you do then you loose the order. So I switch to list to avoid this potential bug.

Reply:
I don't follow how this is relevant. Doesn't this lead to `O(n^2)` though?